[StateMachine] Rework AccountBalances.

batiati opened the pull request:
- Rename `AccountBalances` to `AccountEvents`, inserting an object for every event that updates the balance (transfers and expires), regardless of `Account.flags.history`.

- Add indexes for the account's latest balance update, enabling direct "as-of" balance lookups by timestamp. This index is stored only for accounts with `Account.flags.history`.

- Add indexes for expiry events (dr/cr accounts, pending transfers, and ledgers), providing the same query capabilities for expires as for regular transfers.

- Add indexes for tracking `{post,void}_transfer`, since `Transfers.flags` isn´t indexed.

- Add index for "prunable" events (those not related to accounts marked with the `history` flag), that could be removed once CDC is done.

There are no plans to retrofit the indexes.

batiati committed:
Remove "cache_account_balances" cli arg.
We don't need object cache for account balances.

batiati committed:
[LSM] scan_builder must allow Objects without `id`.

batiati committed:
[StateMachine] Rework AccountBalances

- Rename `AccountBalances` to `AccountEvents`, inserting an object for every event that updates the balance (transfers and expires), regardless of `Account.flags.history`.
- Add indexes for the account's latest balance update, enabling direct "as-of" balance lookups by timestamp. This index is stored only for accounts with `Account.flags.history`.
- Add indexes for expiry events (dr/cr accounts, pending transfers, and ledgers), providing the same query capabilities for expires as for regular transfers.

There are no plans to retrofit the indexes.

batiati committed:
Review
- Add index for "prunable" events, that could be removed once CDC is done.
- Missing index for `Transfers.flags.{post,void}_pending` and expiry events.

batiati committed:
Print object and index sizes in `inspect constants`.
Displays the byte size required to store each StateMachine object and its indexes during the `tigerbeetle inspect constants` command.

batiati committed:
Repurpose reserved bytes to preserve amounts.
- Store both the user-requested amount and the actual recorded amount, as they may differ when posting transfers and balancing accounts.
- Including the amount in history (alongside other transfer-related data such as flags and ledger) can eliminate the need for an additional lookup in certain use cases.
- Since none of these fields are indexed, there is no additional cost incurred.

sentientwaffle commented on src/stdx.zig:
```diff
@@ -785,22 +785,31 @@ fn format_int_size_bin_exact(
     }
 
-    const mags_iec = " KMGTPEZY";
+    const mags_iec = "BKMGTPEZY";
```
```suggestion
    const magnitudes_iec = "BKMGTPEZY";
```

(It took me a moment to decode the old name!)

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -759,7 +889,7 @@ pub fn StateMachineType(
 
             self.prepare_timestamp += switch (operation) {
-                .pulse => 0,
+                .pulse => constants.batch_max.create_transfers, // Max transfers to expire.
```
~Was this a bug before?~ Ah I see -- previously weren't creating any new objects, we were only updating objects in-place. But now we are inserting "account events" with their own timestamps, which means the prepare/commit timestamps are now advanced by transfer expiry!

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -2881,36 +3074,46 @@ pub fn StateMachineType(
+                    .amount_requested = 0,
+                    .amount = p.amount,
+                });
+                self.commit_timestamp = event_timestamp;
```
Since this is safety-critical, we should assert that the value is being increased.

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -305,15 +320,102 @@ pub fn StateMachineType(
+                        }
+                    }.transfer_expired_ledger,
+
+                    // Tracks events for accounts without the history flag,
+                    // enabling a cleanup job to delete them after CDC.
+                    .prunable = struct {
+                        fn prunable(object: *const AccountEvent) ?void {
+                            if (object.dr_account_flags.history or
+                                object.cr_account_flags.history)
+                            {
+                                return null;
+                            } else {
+                                return {};
+                            }
+                        }
+                    }.prunable,
```
:thinking: I was considering whether it makes sense to invert this index, since history=false is the case where we care more about performance anyway. But it is nice that this index exactly corresponds to the entries that can be deleted, so it is fine as-is.

batiati replied:
I tried to balance the load, when `history == false`, we add this `prunnable` index, otherwise we add `account_timestamp`.

sentientwaffle replied:
Thinking about this some more:
- I expect most accounts to have `history == false`,
- which means that this index will be very "dense" -- i.e. it would not be uncommon to have long runs of timestamps that are all in this index,
- whereas if we stored the inverse condition, it would be much more sparse.

TB doesn't support compression right now -- but when it does, the density lends itself well to an inexpensive index (e.g. we could store the first integer in a "run" + "length of run"). So I am okay paying the more expensive cost until then.

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -288,12 +303,12 @@ pub fn StateMachineType(
         };
 
-        const AccountBalancesGroove = GrooveType(
+        const AccountEventsGroove = GrooveType(
```
Since these objects are currently never queried, what can we do to verify that we are indeed inserting the data correctly?

batiati replied:
> Since these objects are currently never queried

We do query `AccountEvents` when calling `get_account_balances`. However, none of the new indexes or fields are used. 

> what can we do to verify that we are indeed inserting the data correctly?

We could assert the data on the write path while it's still in memory (although it's not usually what we do with indexes. There are other cases of never queried indexes for future use). Any thoughts?

sentientwaffle replied:
Hmm, we are handicapped by the fact that the state machine workload doesn't keep a list of all the transfers it has created and instead tries to be clever with workload generation to avoid doing so. I mean to fix that but I doubt I'll get to it soon. :sweat: 

If we did have that, then we could add a temporary checker to the VOPR which randomly fires (perhaps triggered by the replica's `committed` hook)? It would just reach into the replica's forest and run queries and verify that the events match the workload's expectations.

:thinking: Weaker than a fuzz test, but maybe we could check with a state machine unit test? Add a new "action" to our state machine unit test DSL which, when executed, runs a query directly against the state machine's events groove?

batiati committed:
Review

batiati committed:
[StateMachine] Unit tests for `get_events`
- Introduces a temporary, non-exposed `get_events` operation as a shim for testing events. This will be properly implemented (with its VOPR workload) in a follow-up PR.
- Adds unit tests for `get_events`.

batiati committed:
[StateMachine] Assert expiry timestamp during unit tests.

batiati committed:
Tidy: docs.zig: error: 'log' is dead code

batiati committed:
**Review: Fix incorrect `AccountEvent` field order**
- New fields must be added at the `reserved` offset to maintain compatibility with the old `AccountBalance` object.
- Reordered fields across all call sites.
- Renamed the `ledger_expired` index.

sentientwaffle reviewed (approved):
LGTM!
